sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/ProcessBundleHandler.java

```diff
@@ -555,6 +575,38 @@ public BundleFinalizer getBundleFinalizer() {
     }
   }
 
+  private void embedOutboundElementsIfApplicable(
+      ProcessBundleResponse.Builder response, BundleProcessor bundleProcessor) {
+    List<Elements> collectedElements =
+        new ArrayList<>(bundleProcessor.getOutboundAggregators().size());
+    boolean hasFlushedAggregator = false;
+    for (BeamFnDataOutboundAggregator aggregator :
+        bundleProcessor.getOutboundAggregators().values()) {
+      Elements elements = aggregator.sendOrCollectBufferedDataAndFinishOutboundStreams();
+      if (elements == null) {
+        hasFlushedAggregator = true;
+      }
+      collectedElements.add(elements);
+    }
+    if (!hasFlushedAggregator && !collectedElements.isEmpty()) {
+      Elements.Builder elementsToEmbed = Elements.newBuilder();
+      for (Elements collectedElement : collectedElements) {
+        elementsToEmbed.mergeFrom(collectedElement);
+      }
+      response.setElements(elementsToEmbed.build());
+    } else {
+      // If there's flushed aggregator, flush all other aggregators as well.
```

Comment:
```suggestion
      // Since there was at least one flushed aggregator, we have to use the aggregators that were able to successfully collect their elements to emit them and can not send them as part of the ProcessBundleResponse.
```